### PR TITLE
doc: Add documentation for ZC EM Post API

### PR DESCRIPTION
**Pull request by nitbhat**
*Original date: 2019-04-26 21:14:50*
*Original PR: https://charm.cs.illinois.edu/gerrit/5053*

---

This change adds documentation for both p2p and broadcast operations
using the ZC EM Post API.

Change-Id: Ic17133de58a3bf551c3dae17c997390e7f8d4c02

### CODE REVIEW

**Comment by stwhite91** [comment]
The ZC Post API is in 6.10 right? Assuming that, we should target this at 6.10 as well

**Comment by nitbhat** [comment]
Yes, this should be in 6.10.

**Comment by nitbhat** [comment]
Only review the Post API commit. The previous commit is being reviewed in this PR (https://github.com/UIUC-PPL/charm/pull/2157)

File: doc/charm++/manual.rst [hunk 1]
```diff
@@ -4811,55 +4854,187 @@ registration memory handles and will not incur any registration costs.
...
    //called when RDMA operation is completed
-   void zerocopySent(CkDataMsg *m)
+   void zerocopySent(CkDataMsg *msg)
    {
-     //get access to the pointer and free the allocated buffer
-     void *ptr = *((void **)(m->data));
-     free(ptr);
-     delete m;
+     // Cast msg->data to a CkNcpyBuffer to get the source buffer information object
+     CkNcpyBuffer *source = (CkNcpyBuffer *)(m->data));
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst]: should be msg->data I think

File: doc/charm++/manual.rst [hunk 2]
```diff
@@ -4930,6 +5120,127 @@ used.
...
+to receive the data in a user posted buffer, which is allocated and managed by the user.
+
+The posting of the receiver buffer happens at an object level, where each recipient object,
+for example, a chare array element or a group element or nodegroup element posts a receiver
+buffer using a special version of the entry method.
+
+To send an array using the Zero Copy Post API, specify the array
+parameter in the .ci file with the nocopypost specifier.
+
+.. code-block:: c++
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst]: could be charmci code block

File: doc/charm++/manual.rst [hunk 3]
```diff
@@ -4930,6 +5120,127 @@ used.
...
+  void foo (int size, int *arr) {
+    // arr points to myBuffer and the data sent is received in myBuffer
+    // Note that 'arr' buffer is the same as myBuffer, which is user allocated and managed
+
+    computeValues();
+  }
+
+Similarly, for sending and receiving multiple arrays, the .ci file specification will be:
+
+.. code-block:: c++
```
Comment (matthiasdiener) [inline doc/charm++/manual.rst]: could be charmci code block

File: doc/charm++/manual.rst (hunk 1 above)
Comment (nitbhat) [reply doc/charm++/manual.rst]: Good catch. Fixed in this PR (https://github.com/UIUC-PPL/charm/pull/2157)

File: doc/charm++/manual.rst (hunk 2 above)
Comment (nitbhat) [reply doc/charm++/manual.rst]: Done.

File: doc/charm++/manual.rst (hunk 3 above)
Comment (nitbhat) [reply doc/charm++/manual.rst]: Done

File: doc/charm++/manual.rst [hunk 4]
```diff
@@ -4835,7 +4834,6 @@ registered or pinned memory. Registered memory can be de-registered by calling
 the ``deregisterMem()`` method on the CkNcpyBuffer object.
 
 .. code-block:: c++
-
```
Comment (evan-charmworks) [inline doc/charm++/manual.rst]: The removal of this line seems like a typo

Comment (nitbhat) [reply doc/charm++/manual.rst]: Fixed

**Pull request merged**